### DISCUSSION TITLE
Custom Navigator Widget

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#10414] [Q&A] [answered]
Hi,
I'm trying to override `NavigatorWidget` 

custom-widget.tsx
```
import { FileNavigatorWidget } from '@theia/navigator/lib/browser/navigator-widget';
import { TreeNode } from "@theia/core/lib/browser";
@injectable()
export class CustomTree extends FileNavigatorWidget {

    protected handleClickEvent(node: TreeNode | undefined, event: React.MouseEvent<HTMLElement>): void {
        console.log("Works!!);
        super.handleClickEvent(node, event);
    }
}
```

custom-frontend-module.ts
```
    bind(CustomTree).toSelf();
    bind(WidgetFactory).toDynamicValue(({ container }) => ({
        id: 'files',
        createWidget: () => container.get(CustomTree)
    })).inSingletonScope();
```

However, I'm recieving this error. 
![image](https://user-images.githubusercontent.com/80001447/141711339-eaab8ee3-543e-4e91-b534-c9173c70d222.png)

It would be great if you could assist regarding this. 
Thanks

**Comment by msujew** [comment, marked as answer]
Hi @maheshrajrp,

binding a widget related to trees is a bit more complicated compared to our usual services. We create special containers for that use case, which you can see in [this file](https://github.com/eclipse-theia/theia/blob/master/packages/navigator/src/browser/navigator-container.ts).

Luckily, you can use a predefined container for your case:

```ts
function createCustomNavigatorWidget(parent: interfaces.Container): CustomTree {
    const container = createFileNavigatorContainer(parent);
    container.unbind(FileNavigatorWidget);
    container.bind(CustomTree).toSelf();
    return container.get(CustomTree);
}

// and run this in your frontend module
bind(CustomTree).toDynamicValue(ctx =>
    createCustomNavigatorWidget(ctx.container)
);
bind(WidgetFactory).toDynamicValue(({ container }) => ({
    id: FILE_NAVIGATOR_ID,
    createWidget: () => container.get(CustomTree)
})).inSingletonScope();
```

**Reply by maheshrajrp** [reply]
Hi @msujew ,

Thanks !! It works like a charm. Adding to it, I'm trying to change icon and name. What would be the recommended way ?

Thanks

**Reply by msujew** [reply]
You can override the title options that are defined here:

https://github.com/eclipse-theia/theia/blob/c55835835e838afb7d05e11dbfca9517a96f3ec1/packages/navigator/src/browser/navigator-widget-factory.ts#L30-L34

Simply run this code at application startup:

```ts
EXPLORER_VIEW_CONTAINER_TITLE_OPTIONS.label = 'Custom tree';
EXPLORER_VIEW_CONTAINER_TITLE_OPTIONS.iconClass = 'custom-icon'; // The iconClass is a CSS class
```

**Comment by maheshrajrp** [comment]
Hi @msujew ,
This works like a charm. However, I'm curious about this method too. It would be helpful if you could share your thoughts regarding this. I checked in `NavigatorWidgetFactory` and was able to find the location where it was declared. And so I tried to rebind this so as to edit the Labels and also remove the open widget. It didn't work. What am I missing. Can you suggest correct way ? Can I use this method to remove `OpenEditorWidget` ?

Thanks a ton.


Tried to rebind this way. 
 ```
   bind(CustomNavigatorFactory).toSelf().inSingletonScope();
    rebind(NavigatorWidgetFactory).toService(CustomNavigatorFactory);
```

```
export const CUSTOM_EXPLORER_VIEW_CONTAINER_TITLE_OPTIONS: ViewContainerTitleOptions = {
    label: 'File Explorer',
    iconClass: 'volante-file-explorer-icon',
    closeable: false,
};
@injectable()
export class CustomNavigatorFactory extends NavigatorWidgetFactory {

    async createWidget(): Promise<ViewContainer> {
        const viewContainer = this.viewContainerFactory({
            id: EXPLORER_VIEW_CONTAINER_ID,
            progressLocationId: 'explorer'
        });
        viewContainer.setTitleOptions(CUSTOM_EXPLORER_VIEW_CONTAINER_TITLE_OPTIONS);
        const navigatorWidget = await this.widgetManager.getOrCreateWidget(FILE_NAVIGATOR_ID);
        viewContainer.addWidget(navigatorWidget, this.fileNavigatorWidgetOptions); // Not Adding OpenEditorWidget
        return viewContainer;
    }
}
```

**Reply by msujew** [reply]
I think the binding isn't quite correct. The `NavigatorWidgetFactory` is bound to the `WidgetFactory` using something like this:

```ts
bind(WidgetFactory).toService(NavigatorWidgetFactory);
```

You can either do this:

```ts
bind(CustomNavigatorFactory).toSelf().inSingletonScope();
rebind(NavigatorWidgetFactory).to(CustomNavigatorFactory).inSingletonScope();
```

or

```ts
bind(CustomNavigatorFactory).toSelf().inSingletonScope();
bind(WidgetFactory).toService(CustomNavigatorFactory);
```

I'm quite sure one of these should be working (maybe both, I'm not 100% sure though).

**Reply by maheshrajrp** [reply]
Hi @msujew ,

`bind(CustomNavigatorFactory).toSelf().inSingletonScope();
rebind(NavigatorWidgetFactory).to(CustomNavigatorFactory).inSingletonScope();`

 and this worked. Thanks, I finally was able to understand these.
